browser-switch/src/main/java/com/braintreepayments/api/BrowserSwitchRequest.java

```diff
@@ -1,39 +1,58 @@
 package com.braintreepayments.api;
 
 import android.net.Uri;
+import android.util.Base64;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.VisibleForTesting;
 
 import org.json.JSONException;
 import org.json.JSONObject;
 
+import java.nio.charset.StandardCharsets;
+
+// Links
+// Base64 Encode a String in Android: https://stackoverflow.com/a/7360440
+
+// TODO: consider encryption
+// Ref: https://medium.com/fw-engineering/sharedpreferences-and-android-keystore-c4eac3373ac7
+
+// TODO: Rename to `BrowserSwitchStartRequest` and remove `BrowserSwitchOptions` in favor this class
 public class BrowserSwitchRequest {
 
+    private static final String KEY_REQUEST_CODE = "requestCode";
+    private static final String KEY_URL = "url";
+    private static final String KEY_RETURN_URL_SCHEME = "returnUrlScheme";
+    private static final String KEY_METADATA = "metadata";
 
     private final Uri url;
     private final int requestCode;
     private final JSONObject metadata;
     @VisibleForTesting
     final String returnUrlScheme;
-    private boolean shouldNotifyCancellation;
-
-    static BrowserSwitchRequest fromJson(String json) throws JSONException {
-        JSONObject jsonObject = new JSONObject(json);
-        int requestCode = jsonObject.getInt("requestCode");
-        String url = jsonObject.getString("url");
-        String returnUrlScheme = jsonObject.getString("returnUrlScheme");
-        JSONObject metadata = jsonObject.optJSONObject("metadata");
-        boolean shouldNotify = jsonObject.optBoolean("shouldNotify", true);
-        return new BrowserSwitchRequest(requestCode, Uri.parse(url), metadata, returnUrlScheme, shouldNotify);
+
+    @NonNull
+    static BrowserSwitchRequest fromBase64EncodedJSON(@NonNull String base64EncodedRequest) throws BrowserSwitchException {
```

Comment:
I guess we should weigh the benefits of base64encoding compared to the extra merchant integration lift of having to handle the `Failure` type for all browser based flows. I could go either way. What do others think?

Reply:
Oh yeah def. We can't really enforce JSON tampering though. We may still need the failure scenario for the case where we're unable to reconstruct the original `BrowserSwitchRequest`.